Monomorphization of polymorphic maps and binders

shazqadeer opened the pull request:
This PR generalizes the monomorphization feature to handle polymorphic maps, lambdas, and quantifiers. 

A polymorphic map type T is converted into a single-constructor datatype with one field for each instantiation I of T encountered in the program. An operation on a value of I-instantiation of T is translated to the corresponding operation on the field corresponding to I in the datatype corresponding to T.  Polymorphic lambdas are translated to a constructed value of the datatype; the field corresponding to I is initialized by instantiating the polymorphic lambda on I.

A polymorphic forall quantifier is translated to a conjunction of monomorphic instances of the quantifier obtained by instantiating the bound type parameters on concrete types collected from the program. Similarly, a polymorphic exists quantifier is translated to a disjunction. This instantiation scheme is "sound" if each forall quantifier occurs in positive polarity and each exists quantifier occurs in negative polarity in the VC that is checked for satisfiability.

The implementation uses a heuristic for collecting types for monomorphizing quantifiers. The body of the quantifier is examined to collect all instances of polymorphic types and functions mentioned in it. Suppose an instantiation T X of the polymorphic type T (with one parameter instantiated to X) is collected. If X is a type parameter of the enclosing quantifier, then each concrete type used in any instantiation of T becomes a candidate for instantiating the type quantifier. This scheme is easily generalized to types with more than one type parameter. References to polymorphic functions are handled similarly. Once candidates for all bound type parameters of the quantifier are collected, instantiation is performed using the Cartesian product of the candidates for each parameter.

A big source of complexity in the implementation is the handling of nested binders. To handle nesting appropriately, the implementation (1) dynamically discovers polymorphic binders that become candidates for instantiation, and (2) finalizes the instantiation of a binder only when the instantiation of all of its children has been finalized.

gauravpartha reviewed (commented):
Thanks a lot for the PR! It's great to see monomorphization applied to polymorphic maps and binders. Moreover, it's great that the support even works for Dafny- and Viper-generated programs.

I agree with the general approach taken for monomorphizing binders. While going through the PR, I noticed a potential issue for the encoding of polymorphic maps (see comment below) that we should discuss.

Regarding the code:
I did not look at all the details in `Monomorphization.cs`, but I understand what various parts of the code are doing at a high-level.  I have added various comments at different points where I think adding C# comments to the code would help undestanding `Monomorphization.cs`.

I tested the code on simple examples and looked at the monomorphized code (via `/printInstrumented`). In some cases, the code crashed (I have added comments to the lines where crashes occurred with accompanying examples). In many cases, monomorphization did what I expected. In some cases, there were fewer instantiations of axioms than I initially expected (I added comments with examples). 

In addition to the comments added to the lines in the code, I add the following more general remarks.

## Monomorphization of polymorphic maps
I think there is a subtle discrepancy between polymorphic maps and their monomorphized counterparts, which I had not thought about before, which could lead to issues.

Consider the common polymorphic map type `HeapType := <A>[ref, Field A]A` representing a heap. The monomorphized version (which I refer to as `MonoHeapType`) represents this map as a collection of finite maps `[ref, Field A]A` where `A` is instantiated in finitely many ways. 

There is no bijection between `HeapType` and `MonoHeapType` (each element in `MonoHeapType` corresponds to many instances of `HeapType`). As a result, the function `f(h: HeapType) : int` is not fully captured by `fMono(h: MonoHeapType)`. That means, if one has an instantiation for `f`, then one cannot, in general, construct a corresponding instantiation for `fMono`. However, this is what one needs, if one wants to show that correctness of the monomorphic program implies correctness of the polymorphic program.

As a result, I think soundness of monomorphization may not be guaranteed in general when polymorphic maps are involved. One could get around this for Viper by forcing monomorphization to take every relevant instantiation into account. For Dafny, I would be surprised if this were not the case. I think there are alternatives one could take in the Boogie encoding too, but I need to think about those more.

Update: I think one potential alternative that would be easy to implement is the following. Currently, `MonoHeapType` is represented by a datatype with a single constructor, where there is one field per instantiation. One could introduce a fresh type `FreshT` (without type arguments) and add an additional field to this constructor, which represents all other instantiations that monomorphization does not take into account. This field will never be read explicitly. For lambda expressions, one would instantiate that field with an arbitrary value of `FreshT` (via a havoc).

## SMT datatypes
I was surprised that monomorphization works for the following program:

```
  var m : <S>[S]bool;  
 
  assume (forall x : <T>[T]bool :: {m[x]} m[x] <==> !x[x]);
  assert m[m] ==> !m[m];
  assert false;
```
What is interesting here is that to monomorphize the map type `<S>[S]bool`, one must already know `<S>[S]bool`, since the only instantiation is `<S>[S]bool`. 

Monomorphization resolves this by defining a recursive Boogie datatype `type Mono = SomeConstructor ([Mono]bool)`, which is then later translated by Boogie to an SMT datatype. This worried me at first, because the SMTLIB specification requires that datatype definitions are well-founded. This definition shown here is not well-founded because “Mono” occurs in the domain type of a map, which is in a contravariant position. 

However, I don’t think there is a fundamental issue currently, because Boogie does not translate map types to SMT maps and instead introduces new types for SMT sorts to represent the map types.

Note that I am not sure how to give a semantic model for `<S>[S]bool` (or whether there even exists a semantic model). But my remark above suggests that at least that monomorphization is not introducing a fundamental problem (but there may be a fundamental problem for maps of type `<S>[S]bool` in general).

## Representation polymorphic maps with free type variables
I would like to point a discrepancy currently between monomorphization and non-monorphization regarding the representation of polymorphic maps when free type variables are involved, which I think is fine for now (because, as far as I know, in practice the discrepancy does not matter). Consider the following program:

```
function f<T>(x: T) : int;

type Field _;
const f1: Field int;
const f2: Field bool;

procedure p() {
	var y: <A>[Field A]bool;
	var z: <A>[Field A]int;

	assume (forall <T1,T2> x: <A>[Field A]T1, y: <A>[Field A]T2 :: x != y ==> f(x) != f(y));
	assert f(y) != f(z);
}
```
The current monomorphization code registers the following 4 polymorphic map types independently:<A>[Field A]bool, <A>[Field A]int, <A>[Field A]T1, <A>[Field A]T2

Note that the latter two have free type variables (T1 and T2). So, monomorphization does not notice that `<A>[Field A]T1` and `<A>[Field A]T2` actually could coincide with `<A>[Field A]bool` and `<A>[Field A]int` (namely for instantiations T1/T2 := int/bool. Without monomorphization this is taken into account. As far as I know, this discrepancy does not matter for programs generated by Dafny and Viper, so I would not put in the effort to change this.

shazqadeer commented on Test/monomorphize/monomorphize1.bpl:
```diff
@@ -1,4 +1,4 @@
-// RUN: %parallel-boogie -lib:base -monomorphize -useArrayTheory "%s" > "%t"
```
If ```-lib:base``` is used, both ```-monomorphize``` and ```-useArrayTheory``` are turned on.

shazqadeer commented on Source/Core/Monomorphization.cs:
```diff
@@ -181,15 +189,6 @@ private bool IsFinitelyInstantiable()
-
     public override Expr VisitNAryExpr(NAryExpr node)
     {
       if (node.Fun is BinaryOperator op && op.Op == BinaryOperator.Opcode.Subtype)
```
Yes.

shazqadeer commented on Source/Core/Monomorphization.cs:
```diff
@@ -219,102 +217,586 @@ public override Cmd VisitCallCmd(CallCmd node)
+    {
+      if (node.ProxyFor == null)
       {
         isMonomorphizable = false;
```
I added this code a long time ago and may have had an erroneous understanding of TypeProxy at that time. At this time, I don't expect ```node.ProxyFor``` to be null for any well-typed program. Should I delete the override?

gauravpartha replied:
I would either delete the override or add a `Debug.Assert` here.

shazqadeer replied:
I tried to delete the override and discovered that monomorphize/generic-call-arg.bpl fails. Leaving as is.

shazqadeer commented on Source/Core/Monomorphization.cs:
```diff
@@ -611,9 +1152,43 @@ public override Expr VisitLetExpr(LetExpr node)
       return expr;
     }
+
+    private static bool IsPolymorphic(Type type)
+    {
+      return new PolymorphicTypeChecker().IsPolymorphic(type);
+    }
+
+    class PolymorphicTypeChecker
```
The answer to your question is: Yes.

shazqadeer commented on Source/Core/Monomorphization.cs:
```diff
@@ -219,102 +217,586 @@ public override Cmd VisitCallCmd(CallCmd node)
+
+    public FieldAccess GetFieldAccess(List<Type> actualParameters)
+    {
+      Debug.Assert(fieldInfos.ContainsKey(actualParameters));
```
Fixed.

shazqadeer commented on Source/Core/Monomorphization.cs:
```diff
@@ -1166,22 +1788,26 @@ public override Function VisitFunction(Function node)
   public class Monomorphizer
   {
+    /*
+     * This class exposes monomorphization to rest of Boogie. After type checking a Boogie program,
```
Fixed.

shazqadeer commented on Source/Core/Monomorphization.cs:
```diff
@@ -1166,22 +1788,26 @@ public override Function VisitFunction(Function node)
   {
+    /*
+     * This class exposes monomorphization to rest of Boogie. After type checking a Boogie program,
+     * an instance of this class is created and stashed inside the program object.
```
Fixed.

shazqadeer commented on Source/Core/Monomorphization.cs:
```diff
@@ -143,22 +155,18 @@ class MonomorphizableChecker : ReadOnlyVisitor
         .Select(f => f.DefinitionAxiom).ToHashSet();
-      this.axiomsToBeInstantiated = new Dictionary<Axiom, TypeCtorDecl>();
       this.typeVariableDependencyGraph = new Graph<TypeVariable>();
       this.strongDependencyEdges = new HashSet<Tuple<TypeVariable, TypeVariable>>();
```
Fixed.

shazqadeer commented on Source/Core/Monomorphization.cs:
```diff
@@ -219,102 +217,586 @@ public override Cmd VisitCallCmd(CallCmd node)
+     * all instantiations have been discovered. It accomplishes the following substitutions:
+     * - ach access to a polymorphic map with an access to corresponding datatype
+     * - each occurrence of a polymorphic map type with the corresponding datatype
+     * - each polymorphic lambda with a constructor call of corresponding datatype
```
Fixed.

shazqadeer commented on Source/Core/Monomorphization.cs:
```diff
@@ -1135,6 +1752,11 @@ public override AssignLhs VisitFieldAssignLhs(FieldAssignLhs node)
+    }
+
     // this function may be called directly by monomorphizationDuplicator
     // if a non-generic call to a datatype constructor/selector/membership
```
Fixed.

shazqadeer commented on Source/Core/Monomorphization.cs:
```diff
@@ -219,102 +217,586 @@ public override Cmd VisitCallCmd(CallCmd node)
+      var instanceExprsCount = instanceExprs.Count;
+      var typeParameterIndexToTypeHints = Enumerable.Range(0, quantifierExpr.TypeParameters.Count)
+        .Select(_ => new HashSet<Type>()).ToList();
+      for (int typeParamterIndex = 0; typeParamterIndex < quantifierExpr.TypeParameters.Count; typeParamterIndex++)
```
Fixed.

shazqadeer commented on Source/Core/Monomorphization.cs:
```diff
@@ -738,23 +1303,84 @@ class MonomorphizationVisitor : StandardVisitor
     }
 
-    public void AddTriggerType(TypeCtorDecl typeCtorDecl, CtorType ctorType)
+    private void FixpointOnBinderExprMonomorphizers()
```
Done.

shazqadeer commented on Source/Core/Monomorphization.cs:
```diff
@@ -219,102 +217,586 @@ public override Cmd VisitCallCmd(CallCmd node)
+    
+    private MonomorphizationVisitor monomorphizationVisitor;
+    private Dictionary<BinderExpr, BinderExprMonomorphizer> binderExprMonomorphizers;
+    private List<Axiom> splitAxioms;
```
Done.

shazqadeer commented on Source/Core/Monomorphization.cs:
```diff
@@ -990,17 +1626,14 @@ public static string MkInstanceName(string name, List<Type> actualTypeParams)
       if (type is CtorType ctorType && MonomorphismChecker.DoesTypeCtorDeclNeedMonomorphization(ctorType.Decl))
       {
         var instantiatedTypeArguments = ctorType.Arguments.Select(x => LookupType(x)).ToList();
         return new CtorType(Token.NoToken, typeInstantiations[ctorType.Decl][instantiatedTypeArguments],
```
Fixed. See monomorphize/polymap2.bpl.

shazqadeer commented:
@gauravpartha : Regarding your point about polymorphic maps with free type variables, the code had a bug. I have fixed the bug and added your sample as monomorphize/polymap1.bpl.

shazqadeer commented:
@gauravpartha : Regarding your point about the potential unsoundness with monomorphization of polymorphic maps, I do see that if T is a polymorphic map type and Mono(T) is type obtained by monomorphizing T, then the universe of values in T corresponds to a subset of the universe of values in T. Beyond that, I don't have much clarity. To talk precisely about soundness, we would have to come up with a formalization of some core of Boogie and some proof strategy potentially based on some restrictions on the formalized core. It sounds like you have some strategy in mind. Let us discuss.

shazqadeer commented:
@gauravpartha : Regarding the issue of encoding polymorphic maps into well-founded datatypes, your concern is justified.

> However, I don’t think there is a fundamental issue currently, because Boogie does not translate map types to SMT maps and instead introduces new types for SMT sorts to represent the map types.

This is incorrect. Boogie supports translating monomorphic maps to either select-update axioms or to SMT map theory depending on the option ```/useArrayTheory```. Dafny and Viper do not use this option but Civl does. It is important to generate legal SMT2 in both cases.

Do you think it is possible to address your concern by adjusting MonomorphizableChecker and disallowing monomorphization of programs that could create ill-typed SMT2?

shazqadeer commented on Source/Core/Monomorphization.cs:
```diff
@@ -76,6 +79,13 @@ public override Declaration VisitTypeCtorDecl(TypeCtorDecl node)
   {
+    /*
+     * This visitor walks over a type T and adds an edge from any type variable nested inside T
+     * to the type variable "formal". The class MonomorphizableChecker uses this visitor to check
```
Fixed.

shazqadeer commented on Source/Core/Monomorphization.cs:
```diff
@@ -611,9 +1152,43 @@ public override Expr VisitLetExpr(LetExpr node)
+  class MonomorphizationVisitor : VarDeclOnceStandardVisitor
   {
     /*
      * This class monomorphizes a Boogie program. Monomorphization starts from
```
Add the PR summary to that comment.

shazqadeer commented on Source/Core/Monomorphization.cs:
```diff
@@ -219,102 +217,586 @@ public override Cmd VisitCallCmd(CallCmd node)
+    }
+  }
+
+  class InstantiationHintCollector : ReadOnlyVisitor
```
Yes, that is the expected behavior of the current instantiation heuristics. 
- Instantiations [int]bool and [bool]bool for parameter T of f are discovered.
- Inside the assume, the actual parameters for the two calls to f are [T1]bool and [T2]bool.
- The current implementation does not try to match the types [T1]bool and [T2]bool to the known instantiations of f.

On the other hand, suppose we changed the signature of f to 
```
function f<T>(x: [T]bool) : int;
```
Then I would expect the current implementation to discover the appropriate matches: ```{T1, T2} -> {int, bool}```.

To do the right thing on your program, we would have to implement something like e-matching on types.

gauravpartha replied:
> When I step through the code, I see that the quantifier presented to the monomorphization code has both T and R in the outer scope! 

That is indeed surprising. However, your explanation why no instantations are found in both examples makes sense to me.

shazqadeer commented on Source/Core/Monomorphization.cs:
```diff
@@ -181,15 +189,6 @@ private bool IsFinitelyInstantiable()
-      return base.VisitBinderExpr(node);
-    }
-
     public override Expr VisitNAryExpr(NAryExpr node)
```
A function may have a body so could contribute to an expanding type cycle. On the other hand, Select and Store are primitives with no bodies. So I thought it would be safe to omit adding any information about them. Does that make sense?

gauravpartha commented:
> @gauravpartha : Regarding your point about the potential unsoundness with monomorphization of polymorphic maps, I do see that if T is a polymorphic map type and Mono(T) is type obtained by monomorphizing T, then the universe of values in T corresponds to a subset of the universe of values in T. Beyond that, I don't have much clarity. To talk precisely about soundness, we would have to come up with a formalization of some core of Boogie and some proof strategy potentially based on some restrictions on the formalized core. It sounds like you have some strategy in mind. Let us discuss.

The example shown below illustrates why the current monomorphization approach is unsound. I think there is a way of extending the current approach to solve this issue, as I show after explaining the example.

## The problem
Consider the following Boogie program:

```boogie
type Ref;
type Field _;

function P(h: <A> [Ref, Field A]A) : bool;

var h: <A> [Ref, Field A]A;

procedure abstractMethod();
modifies h;
/* If one uncomments this postcondition, then the assertion below should still fail even if
 * extensionality of maps holds (e.g., when using /useArrayTheory or when adding an explicit
 * extensionality axiom) */
// ensures (forall r: Ref, fieldInt: Field int :: h[r, fieldInt] == old(h)[r, fieldInt]);

procedure concreteMethod()
modifies h;
{
    var hBeforeCall: <A> [Ref, Field A]A;
    var hAfterCall: <A> [Ref, Field A]A;

    assume P(h);
    
    hBeforeCall := h;
    call abstractMethod();
    hAfterCall := h;

    /* This assertion should fail.  */
    assert hBeforeCall == hAfterCall;

    /** This assertion holds iff hBeforeCall == hAfterCall holds. */
    assert P(h);
}
```
Ignore the commented-out postcondition of `abstractMethod` for now.

With the current monomorphization approach, the above example verifies even though it should not. The reason why the current approach verifies the above example is: Monomorphization never instantiates the type parameter in `<A>[Ref, Field A]A` (as expected, since there is no corresponding map lookup or map store). As a result, `<A>[Ref, Field A]A`  is monomorphized to a datatype with a single nullary constructor: 

```boogie
type {:datatype} PolymorphicMapType_122;
function {:constructor} PolymorphicMapType_122() : PolymorphicMapType_122;
```

Therefore, according to monomorphization, `<A>[Ref, Field A]A` has only a single value (namely `PolymorphicMapType_122()`). This is an unsound assumption with which one can prove that `abstractMethod` does not modify the global variable `h` even though `abstractMethod`'s contract states the method may modify `h`. `abstractMethod` has no body and Boogie's verification result should apply to any feasible body of `abstractMethod()` (including a body that may change `h`).

The shown problem does not just occur if the monomorphization approach finds no instantiations for the type variable `A` in the polymorphic map. For example, consider the same program after uncommenting `abstractMethods`'s postcondition, which states that `h[r, fieldInt]` is pointwise the same before and after the call for `r: Ref` and `fieldInt: Field int`. Now, monomorphization finds one instantiation, namely `A = int` and represents `<A>[Ref, Field A]A` as a datatype that holds just the map `[Ref,Field int]int` (`Field int` is represented by `Field_41`):

```boogie
type {:datatype} PolymorphicMapType_166;
function {:constructor} PolymorphicMapType_166(PolymorphicMapType_166_41: [Ref,Field_41]int) : PolymorphicMapType_166;
```

This is again an unsound representation, because it assumes that `<A>[Ref, Field A]A` is isomorphic to `[Ref,Field_41]int`, which is not true. In particular, the body of `abstractMethod` may be `h[r,fBool] := false` where `fBool` is of type `Field bool`. 
Using this unsound representation, one can again prove that `abstractMethod` did not change the heap if `[Ref,Field_41]int` is extensional, because of `abstractMethod`'s postcondition. Extensionality can be enforced by using the option `/useArrayTheory` (the example indeed verifies then) or by adding an additional axiom (due to triggering issues, the example does not verify in practice).

Finally, note that the above example is not specific to method calls. Boogie front-ends may represent some arbitrary change of the global variable `h` as an explicit `havoc h` instead of using a method call that may modify `h`. Moreover, to have the desired correspondence between uninterpreted functions `function f(<A>[Ref, Field A]A): ...` and the monomorphized version, one needs to be able to express the monomorphized version in terms of `f`, which is not possible given the current representation.

## A potential solution
The key problem illustrated in the example is that monomorphization essentially assumes that polymorphic maps can be precisely represented as a finite collection of monomorphic maps where the finite collection is obtained by looking at what concrete instances occur in the program. To solve this issue, I think one could do the following.

Let `<A>[K(A)]V(A)` be some polymorphic map with a single type parameter `A` (the solution generalizes easily to multiple parameters). Currently, monomorphization finds some finite instantiations `I1`,`I2`,..., `In` of `A`and then represents `<A>[K(A)]V(A)` as 

```boogie
type {:datatype} PolymorphicMapType_1;
function {:constructor} PolymorphicMapType_1_Constructor(
              Field1: [K(I1)]V(I1), Field2: [K(I2)]V(I2), ..., Fieldn: [K(In)]V(In) ) : PolymorphicMapType_1;
```

Instead of doing this, one could instead represent `<A>[K(A)]V(A)` as:

```boogie
type {:datatype} PolymorphicMapType_1;
type RemainingInstantiationsType_1;
function {:constructor} PolymorphicMapType_1_Constructor(
                Field1: [K(I1)]V(I1), Field2: [K(I2)]V(I2), ..., Fieldn: [K(In)]V(In), 
                RemainingFields: RemainingInstantiationsType_1
              ) : PolymorphicMapType_1;
```

This is identical to the previous representation except that the constructor has one additional field of type `RemainingInstantiationsType_1`, where `RemainingInstantiationsType_1` is some fresh uninterpreted type.  This additional field represents all other instantiations that are not captured by `I1`, `I2`,..., `In`. The field will never be explicitly looked up or modified in the given Boogie program.

An explanation for polymorphic maps is then: A polymorphic map value is a value that collects values for all monomorphic instantiations of the polymorphic map's type parameter. Monomorphization makes a finite number of these instantiations explicit (namely all those that are explicitly accessed in the monomorphized program) and represents the remaining ones using an uninterpreted type.

Using this new representation, one cannot prove that `abstractMethod` does not modify `h` in the above example.

### Lambda expressions with the new representation
With this new representation one must still find a solution for lambda expressions. Currently, `lambda <A> x: K(A) :: e(A)` is represented by `PolymorphicMapType_1_Constructor(lambda x: K(I1) :: e(I1), ..., lambda x: K(In) :: e(In))`. With the new representation one could instead represent the lambda expression as:

```boogie
function lambdaFresh() :: RemainingInstantiationsType_1
PolymorphicMapType_1_Constructor(lambda x: K(I1) :: e(I1), ..., lambda x: K(In) :: e(In),  lambdaFresh())
```

That is, one creates a fresh value for each lambda that represents the value for the remaining instantiations. This way one definitely captures also the value that actually corresponds to the lambda (and many more, but I suspect this won't lead to incompletenesses in practice because the program never looks up the remaining instantiations).

shazqadeer commented:
In your example, you have a function declaration:
```
function P(h: <A> [Ref, Field A]A) : bool;
```

You also have the following sentence:

> Moreover, to have the desired correspondence between uninterpreted functions function f(<A>[Ref, Field A]A): ... and the monomorphized version, one needs to be able to express the monomorphized version in terms of f, which is not possible given the current representation.

I did not understand the issue with functions.

shazqadeer commented:
I understand the modifies issue you are pointing out. It does not seem surprising to me and I am not yet convinced it needs to be fixed. Monomorphization is fundamentally a whole-program analysis. I would think that modularity of the sort you are talking about is an unreasonable expectation. For example, suppose the implementation of a procedure is hidden and the program is translated. Later the procedure is revealed. There is no way to patch the existing translation by augmenting with the translation of the freshly-revealed code. We have to translate the entire program all over again. We have to accept this lack of modularity as the cost of using monomorphization.

shazqadeer commented:
@RustanLeino : FYI

gauravpartha reviewed (approved):
Shaz and I discussed the pending issues. We both agree that one should think more about (1) how (and whether) to achieve that "monomorphized program verifies ==> polymorphic program verifies" (in particular, for the current strategy of polymorphic maps, this relation does not hold --> removing an axiom in the monomorphized program may turn verification failure to verification success since less instances of the polymorphic map are found), (2) how to properly reject programs that could lead to monomorphization producing SMT programs that contain datatypes with definitions that are not well-founded. 

However, it does not make sense to resolve those issues in this PR. Thus, I approve the PR (all my other concerns were handled by Shaz).
